### PR TITLE
Tock `Hello World` demo (alt)

### PR DESCRIPTION
**Pull request by cfrantz**
Provide dependencies on tock and libtock in a way that supports software releases and local development.  Previously, we tried to use `crate_universe` to provide the tock/libtock deps, but `crate_universe` does not allow depending on an arbitrary github archive or a local filepath.  We need to depend on arbitrary github archives because the tock and libtock crates are not available in a crate index (like crates.io).  `crate_universe` does support depending on arbitrary git commits, however, the OpenTitan repo supports building in airgapped mode and bazel does not support arbitrary git commits for airgapped builds.

In order to support arbitrary archives and local development paths, we create a repository rules that permit download+extract or symlinking to a local directory tree.  The rule also supports adding additional files into the target bazel repository to build a hierarchy of BUILD targets.  We also provide a reasonable `crate_build` macro to construct BUILD files with a reasonable `rust_library` rule.

1. Add a `symlink_tree` utility to create a symlinked clone of a local path.
2. Add a `bare_repository` rule that can either download and archive or link to a local subdirectory and insert a user-specified BUILD hierarchy into the repository.
3. Add a repository for `lowRISC/tock` and construct an appropriate BUILD hierarchy.
4. Add a repository for `tock/libtock-rs` and construct an appopriate BUILD hierarchy.

See `//third_party/tock/repos.bzl` to see the external dependency configuration for tock and libtock.

### CODE REVIEW

File: sw/device/silicon_owner/tock/kernel/layout.ld [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        . = . + 384; /* signature */
+        . = . + 4;   /* usage_constraints.selector_bits */
+        . = . + 32;  /* usage_constraints.device_id */
+        . = . + 4;   /* usage_constraints.manuf_state_creator */
+        . = . + 4;   /* usage_constraints.manuf_state_owner */
+        . = . + 4;   /* usage_constraints.life_cycle_state */
+        . = . + 384; /* modulus */
+        . = . + 4;   /* address_translation */
+        . = . + 4;   /* identifier */
+        . = . + 4;   /* signed_region_end */
```
Comment (semihalf-mazurek-michal) [inline sw/device/silicon_owner/tock/kernel/layout.ld:32]: Missing manifest version filed

**Comment by semihalf-mazurek-michal** [comment]
After updating linker script i was able to use custom TockOS/libtock-rs sources and run hello app without complication. 
```
I00002 test_rom.c:203] Boot strap requested
I00000 test_rom.c:160] kChipInfo: scm_revision=eb4b2214
I00001 test_rom.c:186] TestROM:632e99df
I00002 test_rom.c:243] Test ROM complete, jumping to flash (addr: 20000410)!
Unable to find otbn-rsa, disabling RSA support
OpenTitan initialisation complete. Entering main loop
Hello world!
```

File: sw/device/silicon_owner/tock/kernel/layout.ld (hunk 1 above)
Comment (cfrantz) [reply sw/device/silicon_owner/tock/kernel/layout.ld:38]: Fixed.

**Comment by jwnrt** [comment]
Hi @cfrantz, I haven't looked over your PR but had a general question about Tock in the opentitan repo.

Is the intention that we will be doing Tock development (on the kernel and software running on Tock) in this monorepo, or is this work just so that there's a demo in here?

Thanks

File: sw/device/silicon_owner/tock/kernel/BUILD [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        "src/main.rs",
+    ],
+    rustc_flags = [
+        "-g",
+        #        "-Clinker=rust-lld",
+        #        "-Clinker-flavor=ld.lld",
+        #        "-Crelocation-model=static",
+        #        "-Clink-arg=-nmagic",
+        #        "-Clink-arg=-icf=all",
+        #        "-Cforce-frame-pointers=no",
```
Comment (jrvanwhy) [inline sw/device/silicon_owner/tock/kernel/BUILD:45]: Is leaving these commented out intentional?

**Comment by cfrantz** [comment]
Superseded by #19439, #19442, #19477, #19478, #19495.

**Pull request closed without merging**